tools/test_full_sync.py

```diff
@@ -16,7 +21,60 @@
 from chia.cmds.init_funcs import chia_init
 
 
-async def run_sync_test(file: Path, db_version=2) -> None:
+exit_with_failure = False
+
+
+class ExitOnError(logging.Handler):
+    def handle(self, record):
+        pass
+
+    def emit(self, record):
+        if record.level != logging.ERROR:
+            return
+        global exit_with_failure
+        exit_with_failure = True
+
+    def createLock(self):
+        self.lock = None
+
+    def _at_fork_reinit(self):
+        pass
+
+
+@contextmanager
+def enable_profiler(profile: bool, counter: int) -> Iterator[None]:
+    if profile:
+        pr = cProfile.Profile()
+        pr.enable()
+        receive_start_time = time.monotonic()
+        try:
+            yield
+            pr.disable()
+            if time.monotonic() - receive_start_time > 10:
+                pr.create_stats()
+                pr.dump_stats(f"slow-batch-{counter:05d}.profile")
+        except:  # noqa: E722
+            pr.disable()
+            raise
+    else:
+        yield
```

Comment:
Since I'm already commenting on other things...  I'll go for this too.  Less indentation and I find that putting the short case first, especially when it just bypasses the entire function, makes it a bit more clear.  Also, put as little as you need in the `try:`.  If possible, have common paths in `finally:`.  Bonus, no more `noqa:`.  Except then I saw the profiler can just be a context manager...  so I deleted all the `finally:` style.

```suggestion
    if not profile:
        yield
        return

    with cProfile.Profile() as pr:
	receive_start_time = time.monotonic()
        yield

    if time.monotonic() - receive_start_time > 10:
        pr.create_stats()
        pr.dump_stats(f"slow-batch-{counter:05d}.profile")
```

Reply:
one question I couldn't find the answer to though; if I want to perform some work only in the failing path, if I can't use `except:` what do I do instead?

Reply:
If you want to respond to errors coming out of the `yield`, you can put a normal `try`/`except` back around it for your needs.  It is still good though to use context managers for the things that provide them.  One aspect is that then you leave the shutdown path in the library's control.  If they decide they need to do something differently they can make the change and you will get it without having to change your code.  Also, consider the two cases below.

```python
try:
    yield
finally:
    shutdown_x()
    shutdown_y()
```

vs.

```python
with Y() as y:
    with X() as x:
        yield
```

The latter will always try to shutdown 'y'.  The former will only try to shutdown 'y' if shutting down 'x' doesn't fail.  So even when a context manager implementation is trivial, using it results in code structure that is much more robust and also more maintainable.

It isn't applicable here, but while I'm going on about context managers, one of the questions that comes up is 'how do i enter some context managers that are not known until run time?'  Or 'how do I conditionally enter a context manager?' (which could be used here) 
 That's what [`contextlib.ExitStack`](https://docs.python.org/3.10/library/contextlib.html#contextlib.ExitStack) and [`contextlib.AsyncExitStack`](https://docs.python.org/3.10/library/contextlib.html#contextlib.AsyncExitStack) are for.  Hopefully these senseless snippets are sufficient to show the form, even if not the use case.

```python
def f(list_of_context_managers_to_use):
    with contextlib.ExitStack() as exit_stack:
        for manager in list_of_context_managers:
            exit_stack.enter_context(manager)
        do_something_with_the_managers_having_been_entered()
```

or

```python
with contextlib.ExitStack() as exit_stack:
    if need_manager:
        exit_stack.enter_context(manager)
    do_something_with_the_manager_maybe_having_been_entered()
```

I'm thinking that lack of awareness of exit stacks might relate to the async generators that we have for our test setup.  Though I guess they could be statically nested as well.  Anyways, context managers for the win basically anytime you have a resource and want to do something with it when the rest of the code is done with it.  And then we hit the structured concurrency patterns that makes it possible to use context managers in lots of places we presently can't due to how we have structured our tasks.

https://github.com/Chia-Network/chia-blockchain/blob/e47958605a8ed360f07591d14b59a468f26f00ad/tests/setup_nodes.py#L339-L369

Reply:
I see. that can also help reducing indentation levels, even when all the context managers are known statically.